### DISCUSSION TITLE
[mini.pick] Project Picker

### DISCUSSION THREAD

**Discussion by wroyca** [nvim-mini/mini.nvim#1141] [Show and tell]
Here's a tiny project picker that I thought I'd share. It's "hardcoded" to use ~/Projects for performance. Keep in mind that it hasn't been thoroughly tested and was mainly an ad hoc prototype for my own use :)

```lua
local projects = function(_, opts)
  local projects = {}
  local projects_dir = vim.fn.expand("~/Projects")

  for name, type in vim.fs.dir(projects_dir) do
    if type == "directory" then
      table.insert(projects, { text = name, dir = projects_dir .. '/' .. name })
    end
  end

  local choose = function(item)
    if not item then return end
    vim.fn.chdir(item.dir)

    -- Schedule to focus mini.files, as it doesn't otherwise.
    vim.schedule(function()
      require("mini.files").open(item.dir)
    end)
  end

  local preview = function(buf_id, item)
    if not item then return end

    local files = {}
    for name, type in vim.fs.dir(item.dir) do
      if type == "file" and not name:match("^%.") then
        table.insert(files, name)
      end
    end

    vim.api.nvim_buf_set_lines(buf_id, 0, -1, false, files)
  end

  local source = { items = projects, name = "Projects", choose = choose, preview = preview }
  opts = vim.tbl_deep_extend("force", { source = source }, opts or {})
  return MiniPick.start(opts)
end
```

![Screenshot from 2024-08-12 22-07-53](https://github.com/user-attachments/assets/3a46a72e-f86f-4d35-8880-41a00dc59438)

![image](https://github.com/user-attachments/assets/a3a6204c-a109-476b-b592-88f45f9a6cb8)

![image](https://github.com/user-attachments/assets/6748f2a7-bcf7-4b03-8a42-f9e3bc05fbba)

**Comment by echasnovski** [comment]
Nice! This can be simplified, though:
- Instead of using CLI `find`, I'd recommend using `vim.fn.readdir()` (possibly together with `vim.fn.isdirectory()`). Or even `vim.fs.dir()` which already returns an iterator over entry name and its type (file, directory, etc.).
- `vim.cmd.cd(item.dir)` can be `vim.fn.chdir(item.dir)`. There are tiny differences (like the latter will automatically set the "most local" working directory), but otherwise they are the same. It's just usually cleaner to use function directly (if it exists) than command through `vim.cmd`.
- I am not really sure about motivation behind showing only files in preview. I'd say that just listing directory contents is more universally useful.

**Reply by wroyca** [reply]
> Instead of using CLI find, I'd recommend using vim.fn.readdir() (possibly together with vim.fn.isdirectory()). Or even vim.fs.dir() which already returns an iterator over entry name and its type (file, directory, etc.).

TIL, will take a look at some point. Thanks!

> vim.cmd.cd(item.dir) can be vim.fn.chdir(item.dir). There are tiny differences (like the latter will automatically set the "most local" working directory), but otherwise they are the same. It's just usually cleaner to use function directly (if it exists) than command through vim.cmd.

Figured there'd be an API for that. Couldn't find it yesterday though, and it was getting late. Thanks!

> I am not really sure about motivation behind showing only files in preview. I'd say that just listing directory contents is more universally useful.

Haha, yes. I wasn't sure what to put as the preview but figured that something was better than nothing, and it feels consistent (with other pickers) to quickly have a view of it. But yeah, there's no profound rationale here—it's optional. :)

**Reply by wroyca** [reply]
@echasnovski Added your recommendations; let me know if there's more improvement that could be done. :)

**Reply by echasnovski** [reply]
> @echasnovski Added your recommendations; let me know if there's more improvement that could be done. :)

Since you asked ( :) ), I'd personally wouldn't use `vim.fn.fnamemodify(item.dir .. '/' .. name, ":~:s?\\~/Projects/??")`. The `string.sub()` approach looks better (more explicit) here: `string.sub(item.dir .. '/' .. name, projects_dir:len() + 2)`. Or even just `name`.

By the way, I also realized (during trying this out, since I have the same kind directory in '~/repos'), that this is can be implemented as a wrapper for `MiniExtra.pickers.explorer()`. Here is what I'd use:

```lua
  MiniPick.registry.projects = function()
    local cwd = vim.fn.expand('~/Projects')
    local filter = function(data) return data.fs_type == 'directory' end
    local choose = function(item)
      vim.fn.chdir(item.path)
      vim.schedule(function() MiniFiles.open() end)
    end

    return MiniExtra.pickers.explorer({ cwd = cwd, filter = filter }, { source = { choose = choose } })
  end
```

**Reply by wroyca** [reply]
> > @echasnovski Added your recommendations; let me know if there's more improvement that could be done. :)
> 
> Since you asked ( :) ), I'd personally wouldn't use `vim.fn.fnamemodify(item.dir .. '/' .. name, ":~:s?\\~/Projects/??")`. The `string.sub()` approach looks better (more explicit) here: `string.sub(item.dir .. '/' .. name, projects_dir:len() + 2)`. Or even just `name`.
> 
> By the way, I also realized (during trying this out, since I have the same kind directory in '~/repos'), that this is can be implemented as a wrapper for `MiniExtra.pickers.explorer()`. Here is what I'd use:
> 
> ```lua
>   MiniPick.registry.projects = function()
>     local cwd = vim.fn.expand('~/Projects')
>     local filter = function(data) return data.fs_type == 'directory' end
>     local choose = function(item)
>       vim.fn.chdir(item.path)
>       vim.schedule(function() MiniFiles.open() end)
>     end
> 
>     return MiniExtra.pickers.explorer({ cwd = cwd, filter = filter }, { source = { choose = choose } })
>   end
> ```

> By the way, I also realized (during trying this out, since I have the same kind directory in '~/repos'), that this is can be implemented as a wrapper for MiniExtra.pickers.explorer(). Here is what I'd use:

Brilliant, perhaps new Projects picker in mini.extra? ;) 

> string.sub(item.dir .. '/' .. name, projects_dir:len() + 2). Or even just name.

uh neat, will check it out. Thanks!

**Reply by echasnovski** [reply]
> Brilliant, perhaps new Projects picker in mini.extra? ;)

No, I don't think so :sweat_smile: But I *think* I'll add the version of this to my 'init.lua':

```lua
MiniPick.registry.projects = function()
  local cwd = vim.fn.expand('~/repos')
  local choose = function(item)
    vim.schedule(function() MiniPick.builtin.files(nil, { source = { cwd = item.path } }) end)
  end
  return MiniExtra.pickers.explorer({ cwd = cwd }, { source = { choose = choose } })
end
```

**Reply by wroyca** [reply]
> > Brilliant, perhaps new Projects picker in mini.extra? ;)
> 
> No, I don't think so 😅 But I _think_ I'll add the version of this to my 'init.lua':
> 
> ```lua
> MiniPick.registry.projects = function()
>   local cwd = vim.fn.expand('~/repos')
>   local choose = function(item)
>     vim.schedule(function() MiniPick.builtin.files(nil, { source = { cwd = item.path } }) end)
>   end
>   return MiniExtra.pickers.explorer({ cwd = cwd }, { source = { choose = choose } })
> end
> ```
I'm still experimenting, but let me share my thought process. I usually start Neovim in the individual project directory. When I need to switch to another project, it's likely because I need to work on it directly, so having mini.files for renaming, editing, or moving files is necessary. If I want to work on specific files, I can simply open them as needed. Wrapping Explorer works more or less the same way, but you lose the direct flexibility—though you can mimic it to an extent with bindings, I believe. :)

**Reply by echasnovski** [reply]
> When I need to switch to another project, it's likely because I need to work on it directly, so having mini.files for renaming, editing, or moving files is necessary.

I tend to have both "work on the project" and "open file from that project for reference". The first one is covered by sessions and `MiniSessions.select()`, i.e. if there is "work on the project" then there should be session for it. The second one is also quite frequent (like take a look at specific Neovim source code file, for example) and which this new "projects" picker is intended to remedy. Up until now using `:Pick visit_paths` was enough (and even more straightforward as there is no need to pick the project first), but I'll try this one instead.

**Reply by evantravers** [reply]
I came here looking for this… I wanted to be able to access my folder of notes anywhere on my system. 

It makes me kind of wish that `:Pick files` could take a `path=` option to cover options like this.

**Reply by echasnovski** [reply]
You can make it accept `cwd` option, which is the same effect. There is [an example just for that](https://github.com/echasnovski/mini.nvim/blob/848c5e8f428faf843051768e0d56104cd02aea1f/doc/mini-pick.txt#L1204-L1209).

The reason this is not built-in is because `cwd` is something that any source can have, so in that case it would have made sense to add to every source. But the latter is a bit too much.

**Reply by evantravers** [reply]
It makes a lot of sense. I did read through the docs and was able to scrabble this ugly one liner that does (I believe) the same thing: https://github.com/evantravers/dotfiles/blob/3a33131a6a4dbe042fff09b1622494c557ff6f3a/home-manager/nvim/default.nix#L211

**Reply by echasnovski** [reply]
Yep, it is the same. Using `vim.fn.expand` is unnecessary, though. 'mini.pick' is smart enough to work with '~' at the start of the path.

**Reply by trickyni** [reply]
> > Brilliant, perhaps new Projects picker in mini.extra? ;)
> 
> No, I don't think so 😅 But I _think_ I'll add the version of this to my 'init.lua':
> 
> ```lua
> MiniPick.registry.projects = function()
>   local cwd = vim.fn.expand('~/repos')
>   local choose = function(item)
>     vim.schedule(function() MiniPick.builtin.files(nil, { source = { cwd = item.path } }) end)
>   end
>   return MiniExtra.pickers.explorer({ cwd = cwd }, { source = { choose = choose } })
> end
> ```

could this be integrated into the example section of the docs? I find this incredibly useful.

**Reply by echasnovski** [reply]
This is probably a bit too much for 'mini.pick' or 'mini.extra' docs. Maybe for future wiki (no established timeline though, sorry).

**Reply by trickyni** [reply]
Understood ^^
Regardless, this will now live happily in my init.lua as well :)

**Comment by wroyca** [comment]
If you happen to have a minute to spare, any recommendations on how I could approach adding icons? I did a vague attempt yesterday based on the pick source code and ended up adding the whole plumbing for H.get_icon, and it ... didn’t go well :)

**Reply by echasnovski** [reply]
Where specifically do you want icons? If in "main" item list, then [`MiniExtra.pickers.explorer()` wrapper](https://github.com/echasnovski/mini.nvim/discussions/1141#discussioncomment-10327131) already has it out of the box and I'd suggest just using it.

As a general suggestion for icons I'd recommend to make custom `show` which calls `MiniPick.default_show(buf_id, items, query, { show_icons = true })`. For it to work properly items should be [compatible with common item structure for paths](https://github.com/echasnovski/mini.nvim/blob/cf81e70e8f1fdc36a3c684efc45c4ee561285137/doc/mini-pick.txt#L252). In particular, either be tables with `path` being full path or plain strings (needs properly set `source.cwd` for relative paths to work).

**Reply by wroyca** [reply]
> Where specifically do you want icons?  

directories and files (normal and preview)

> If in "main" item list, then [`MiniExtra.pickers.explorer()` wrapper](https://github.com/echasnovski/mini.nvim/discussions/1141#discussioncomment-10327131) already has it out of the box and I'd suggest just using it.

I did actually give it a fair chance yesterday, but it didn't fit well with my needs. :sweat_smile: 

> { show_icons = true }

Interesting, will check it out. Thanks!
